### PR TITLE
Add TF implementation of LongT5

### LINKED ISSUE #18063: Add TF implementation of LongT5 model
### Feature request

Add TF implementation of LongT5 model

### Motivation

Add support for TF backend to allow using LongT5 with TF.

### Your contribution

I will add this :]

### PR DESCRIPTION
**Pull request by stancld**
Fixes #18063 

Add:
- [ ] Fix PT-TF equivalence (Local)
- [ ] Fix PT-TF equivalence (TGlobal)
- [ ] Run all slow tests
- [ ] Prepare TF checkpoints
  - [long-t5-local-base](https://huggingface.co/Stancld/long-t5-local-base)
  - [long-t5-local-large](https://huggingface.co/Stancld/long-t5-local-large)
  - [long-t5-tglobal-base](https://huggingface.co/Stancld/long-t5-tglobal-base)
  - [long-t5-tglobal-large](https://huggingface.co/Stancld/long-t5-tglobal-large)
  - long-t5-tglobal-xl https://github.com/huggingface/transformers/issues/19965

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by stancld** [comment]
Hi @patrickvonplaten and @gante,
FYI I've been gradually fixing some PT-TF discrepancies -- I should have some spare time again next weekend, so hopefully, then it should be ready for review :]

**Comment by stancld** [comment]
Not yet closed. Will update at the weekend.

**Comment by gante** [comment]
@stancld should I reopen the PR? :)

**Comment by stancld** [comment]
Hi @gante, yes, I'd open that again.

I apologize for being so slow here, but I've been pretty busy now. I'll try to finish this.

**Comment by gante** [comment]
No worries @stancld, take your time 🤗 And thank you for working on it!

**Comment by stancld** [comment]
Hi @gante, I managed to fix some bugs. There are still some discrepancies between PT and TF implementations. Would you mind having a first look if you spot any obvious differences, please? :]

Otherwise, TF-only tests seem to be passing 🐰

**Comment by gante** [comment]
@stancld Will have a look 👍 Can I have a copy of the error(s) you see locally? (I'm assuming on the slow tests)

**Comment by patrickvonplaten** [comment]
Also cc @ArthurZucker here

**Comment by stancld** [comment]
> @stancld Will have a look 👍 Can I have a copy of the error(s) you see locally? (I'm assuming on the slow tests)

Sorry for the late reply. I fail on `PT-TF` equivalence tests, basically, saying there's a too high difference between outputs.

File: src/transformers/models/longt5/modeling_tf_longt5.py [hunk 1]
```diff
@@ -0,0 +1,2363 @@
...
+        # Apply Feed Forward layer
+        hidden_states = self.layer[-1](hidden_states, training=training)
+        outputs = (hidden_states,)
+
+        # Add attentions if we output them
+        outputs = outputs + (present_key_value_state,) + attention_outputs
+        return outputs  # hidden-states, present_key_value_states, (self-attention weights), (self-attention position bias), (cross-attention weights), (cross-attention position bias)
+
+
+####################################################
```
Comment (gante) [inline src/transformers/models/longt5/modeling_tf_longt5.py:1355]:
```suggestion
```
(we don't use this sort of separation in the codebase, by default)

File: src/transformers/models/longt5/modeling_tf_longt5.py [hunk 2]
```diff
@@ -0,0 +1,2363 @@
...
+# The full model without a specific pretrained or finetuning head is
+# provided as a tf.keras.layers.Layer usually called "TFLongT5MainLayer"
+####################################################
```
Comment (gante) [inline src/transformers/models/longt5/modeling_tf_longt5.py:1358]:
```suggestion
```

File: src/transformers/models/longt5/modeling_tf_longt5.py [hunk 3]
```diff
@@ -0,0 +1,2363 @@
...
+            )
+        else:
+            return TFBaseModelOutput(
+                last_hidden_state=hidden_states,
+                hidden_states=all_hidden_states,
+                attentions=all_attentions,
+            )
+
+
+####################################################
```
Comment (gante) [inline src/transformers/models/longt5/modeling_tf_longt5.py:1604]:
```suggestion
```

File: src/transformers/models/longt5/modeling_tf_longt5.py [hunk 4]
```diff
@@ -0,0 +1,2363 @@
...
+# TFLongT5PreTrainedModel is a sub-class of tf.keras.Model
+# which take care of loading and saving pretrained weights
+# and various common utilities.
+# Here you just need to specify a few (self-explanatory)
+# pointers for your model.
+####################################################
```
Comment (gante) [inline src/transformers/models/longt5/modeling_tf_longt5.py:1610]:
```suggestion
```

File: src/transformers/models/longt5/modeling_tf_longt5.py [hunk 5]
```diff
@@ -0,0 +1,2363 @@
...
+    def get_input_embeddings(self):
+        return self.shared
+
+    def set_input_embeddings(self, value):
+        self.shared = value
+        self.encoder.embed_tokens = self.shared
+        if hasattr(self, "decoder"):
+            self.decoder.embed_tokens = self.shared
+
+    def _shift_right(self, input_ids):
```
Comment (gante) [inline src/transformers/models/longt5/modeling_tf_longt5.py:1657]: This one could get a `#copied from` T5, like in the PT version, correct?

**Review by gante — COMMENTED** [review]
@stancld the PR looks 🔥 , couldn't find anything suspicious 🤔 

If the difference is not too large, (<1e-4), I'd say raising the threshold should suffice. If the difference is larger than 1e-4, then we should try to pin the difference, as it would be very valuable information (not necessarily in this PR, we could delegate it to the future with a `TODO`)

**Comment by patrickvonplaten** [comment]
Almost there I think :-)

**Pull request closed without merging**